core/src/main/java/fj/data/PriorityQueue.java

```diff
@@ -0,0 +1,206 @@
+package fj.data;
+
+import fj.Equal;
+import fj.F;
+import fj.Monoid;
+import fj.Ord;
+import fj.P;
+import fj.P2;
+import fj.P3;
+import fj.Show;
+import fj.data.fingertrees.FingerTree;
+
+import static fj.data.List.list;
+
+/**
+ * A priority queue implementation backed by a {@link fj.data.fingertrees.FingerTree}.  The finger tree nodes are annotated with type K, are combined using a monoid of K and both the key and value are stored in the leaf.  Priorities of the same value are returned FIFO (first in, first out).
+ *
+ * Created by MarkPerry on 31 May 16.
+ */
+public class PriorityQueue<K, A> {
+
+    private final FingerTree<K, P2<K, A>> ftree;
+    private final Equal<K> equal;
+
+    private PriorityQueue(Equal<K> e, FingerTree<K, P2<K, A>> ft) {
+        equal = e;
+        ftree = ft;
+    }
+
+    /**
+     * Creates a priority queue from a finger tree.
+     */
+    public static <K, A> PriorityQueue<K, A> priorityQueue(Equal<K> e, FingerTree<K, P2<K, A>> ft) {
+        return new PriorityQueue<K, A>(e, ft);
+    }
+
+    /**
+     * Creates an empty priority queue.
+     *
+     * @param m A monoid to combine node annotations.
+     * @param e A value to compare key equality.
+     */
+    public static <K, A> PriorityQueue<K, A> empty(Monoid<K> m, Equal<K> e) {
+        return priorityQueue(e, FingerTree.empty(m, (P2<K, A> p) -> p._1()));
+    }
+
+    /**
+     * An empty priority queue with integer priorities.
+     */
+    public static <A> PriorityQueue<Integer, A> emptyInt() {
+        return priorityQueue(Equal.intEqual, FingerTree.empty(Monoid.intMaxMonoid, (P2<Integer, A> p) -> p._1()));
```

Comment:
There is a bunch of these where I later added convenience methods and did not go back and improve the original code.  Thanks for picking this up.